py/runtime: Check for null objects coming from native functions.

agatti opened the pull request:
### Summary

When performing inplace binary operations (eg +=, -=, etc) from a native emitted context, the left hand side operand would be NULL if it wasn't defined earlier.  The code assumed the operand would always be non-NULL and hence a crash occurred when accessing the operand object.

This commit turns that crash into an exception being raised, with the same name and message as if the same error occurred in a bytecode context.  Also, the exception is raised at runtime, following the same behaviour as if the function was not decorated as native.

To reproduce the issue just try to execute the following Python code:

```python
@micropython.native
def f():
	a += 10

f()
```

Whilst this is a minor issue, the very fact that it would crash the whole interpreter because of a typo in a native function makes it quite annoying to have...

This fixes issue #15670.

### Testing

This was tested on the Unix port, but the issue would happen on any target that supports native compilation.

### Trade-offs and Alternatives

This patch is only enabled if `MICROPY_ENABLE_COMPILER` is enabled, and should add a hundred bytes or so of code.  Rather than patching the runtime this issue could maybe be fixed when compiling the incorrect statement that emits the binary operation, but I'm quite new to that part of the codebase.  At least it follows the same behaviour as the VM, better than nothing...

agatti commented:
Added test to keep codecov happy...

dpgeorge commented:
Thanks for attempting to fix this.

The reason this hasn't been fixed before is because (1) it's difficult to do properly, and (2) it will lead to a decrease in performance (for either just native code, or both native and bytecode).

There are actually quite a few tests which test for unbound locals, eg `tests/basics/unboundlocal.py`.  All of these tests are currently disabled when the native emitter is used, because they crash it.

For example, how about `x = ~x`, or `print(x)`?

agatti commented:
Thanks for the explanation!

From some random testing I've made the only other place where I was able to reproduce the issue on the REPL is when handling unary operations (like in `x = ~x` as you mentioned) and can be fixed in the same way as binary operations were.  Everything else was handled properly, but I'll take a deeper look at the test cases you mentioned - if I find the time I may even try to see whether a few hours of `afl-fuzz` yield different crash cases.

Regarding the performance hit, maybe adding a NULL check in a couple of places can probably end up too small of a slowdown to be really noticed - those checks can be optimised/moved around by the C compiler, after all.  If performance in those bits still critical even after the compiler did its magic, would putting those extra checks being an additional define work?

agatti commented:
Ok, I've let `afl-fuzz` go wild for a couple of hours using `annotate_var`, `del_deref`, `del_local`, `scope_implicit`, and `unboundlocal` with all functions marked as native and `native_misc` with the extra test I added first, and all it found were crashes in `mp_unary_op`.  Once that was patched too, nothing interesting was found so far but I'll let it go until tomorrow morning to see if it can come up with something else.

Regarding `unboundlocal`, I ran it on a patched interpreter after marking both functions as native, and I saw no crashes.  Is there any other invalid behaviour I should have noticed?

dpgeorge commented:
I'm surprised fuzzing didn't find more cases.  For example the following easily crashes the unix port:
```py
@micropython.native
def f():
    print(x)
    x = 1
f()
```

agatti commented:
> I'm surprised fuzzing didn't find more cases. For example the following easily crashes the unix port:

It started finding a few more when I upgraded AFL++ to the latest stable version, but just two or three more cases.

Anyway, I've cleaned this up and put the whole lot behind its own define, not enabled by default.  Maybe this should be brought in when the `DEBUG` variable is set in the makefile or something like that.

agatti committed:
py/runtime: Check for null objects coming from native functions.

Native emitted functions do not perform any NULL checks before calling
into the MicroPython runtime, and some runtime functions require
incoming objects not not be NULL.

This commit adds NULL checks for incoming objects to some of those
functions, at the expense of a minor performance decrease due to the
extra instructions.  Crashes are effectively turned into exceptions,
following the behaviour of the bytecode VM.

This fixes issue micropython#15670.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

stinos commented on py/runtime.c:
```diff
@@ -285,6 +285,12 @@ void mp_delete_global(qstr qst) {
 
+    #if MICROPY_COMP_EXTRA_VALIDATION
+    if (arg == MP_OBJ_NULL) {
+        goto unbounded_local_variable;
```
Why `goto` instead of a direct call?

stinos commented on py/mpconfig.h:
```diff
@@ -500,6 +500,12 @@
 
+// Whether to enable extra arguments validation when running native code
+// Requires MICROPY_DYNAMIC_COMPILER and costs 160 bytes (Thumb2)
+#ifndef MICROPY_COMP_EXTRA_VALIDATION
```
'COMP' is 'COMPILER' here, right? How does this require MICROPY_DYNAMIC_COMPILER? Still this doesn't seem to describe well what is actually implemented here: adding checks in runtime.c means they get are used whenever running any type of code, be it Python or native or calling the functions directly from C. To me this seems as if it should be just called 'extra argument validation'. But then I'm wondering if this shouldn't be an 'all or nothing' approach. I didn't check in detail, but it seems not unlikely that not all possible cases have been covered yet?

agatti replied:
The rationale is that the issues these checks handle seem to come only from emitted native code, hence the "dependency" on MICROPY_DYNAMIC_COMPILER and why it is disabled by default.  That should be probably be reworded, as you point out.  If you're not using any native emitters or you already have other error checking in place at the source level then you don't need this.  The name has COMP in it because I didn't find any better stanza for the define in the configuration file and I just followed the naming convention.

It is true this may not cover all possible cases, but it handles all the ones I've bumped into plus a few more found by fuzzing after the first @dpgeorge comment.  I'm quite new to that bit of MicroPython internals after all.

dpgeorge commented:
I really don't think this is the right approach.  There are so many combinations / cases of unbound variables being passed in as arguments.  Basically every C function needs to check it's arguments if they are `MP_OBJ_NULL` or not.

stinos commented:
> Basically every C function needs to check it's arguments if they are MP_OBJ_NULL or not.

Just to be clear: is it so that micropython.native emits bytecode and hence essentially all functions which can get called in vm.c would need a check? Which in turn is indeed pretty much everything from the public API?

dpgeorge commented:
> is it so that micropython.native emits bytecode and hence essentially all functions which can get called in vm.c would need a check?

Not exactly.

The point is, if a local variable is not populated in a `@micropython.native` function then it has an undefined value (probably zero'd out to `MP_OBJ_NULL`), and that null object can be passed anywhere a normal object may be passed, via native calls to functions in `mp_fun_table`.

So to cover every case I think it's sufficient if every function in `mp_fun_table` checks it's arguments for non-null.

stinos commented:
> every function in mp_fun_table

Ok thanks, that's what I wanted to know.

agatti commented:
> I really don't think this is the right approach. There are so many combinations / cases of unbound variables being passed in as arguments. Basically every C function needs to check it's arguments if they are `MP_OBJ_NULL` or not.

Fair point, but I can only think of two other ways to solve this, and I suspect both of those have already been considered before.

The first one is to emit the NULL check along with the native opcodes, as the number and types of arguments for each function in `mp_fun_table` is known beforehand and can be looked up from a table somehow, the other one would be to perform checks at the level before the actual opcodes are emitted - for the latter case I can't say I know where those bits should go or how they should look like, though.

I'd say the first one is slightly more approachable, but maybe it's because I've recently worked with native emitters.  It may very well be I'm missing something important to invalidate that assumption.

So, is there any realistic way to sort this out?

dpgeorge commented:
Just to follow up with some more examples that don't work with the PR in it's current state:
```py
@micropython.native
def f():
    x = 0
    print(x, y)
    y = 1
f()
```
and
```py
@micropython.native
def f():
    l = []
    l.insert(0, x)
    print(l)
    x = 0
f()
```
Both those examples crash the unix port with this PR and `MICROPY_COMP_EXTRA_VALIDATION=1`.

> So to cover every case I think it's sufficient if every function in `mp_fun_table` checks it's arguments for non-null.

Given the `list.insert` example above, that original statement of mine is wrong.  Basically every single native function needs to have a NULL check for all of its arguments.

> So, is there any realistic way to sort this out?

I don't think so.  All known ways to fix this are difficult:
1. put NULL checks everywhere, for all input arguments to all C functions
2. put NULL checks in the code generated by the native emitter
3. do compile time analysis to work out which variables may be accessed before being assigned
